src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,145 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            Node<T> first = new Node<>(null, value, null);
+            head = first;
+            tail = first;
+        } else {
+            Node<T> last = tail;
+            Node<T> newNode = new Node<>(last, value, null);
+            last.next = newNode;
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 0 || index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> left = current.prev;
+            Node<T> newNode = new Node<>(left, value, current);
+            if (left == null) {
+                head = newNode;
+            } else {
+                current.prev = newNode;
+                left.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecker(index);
+        return getNode(index).items;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexChecker(index);
+        T oldValue = getNode(index).items;
+        getNode(index).items = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecker(index);
+        Node<T> oldNode = getNode(index);
+        Node<T> prevNode = oldNode.prev;
+        Node<T> nextNode = oldNode.next;
+        final T returnElement = oldNode.items;
+        if (prevNode == null) {
+            head = nextNode;
+        } else {
+            prevNode.next = nextNode;
+        }
+        if (nextNode == null) {
+            tail = prevNode;
+        } else {
+            nextNode.prev = prevNode;
+        }
+        size--;
+        return returnElement;
```

Comment:
let's move this logic to the separate method delete/unlink and call this method in the method remove by index and remove by value

Reply:
I moved this logic in new method, but I can't understand why, I call this method in remove by object and we can see realisation this method remove by index method. Now I have new method deletByIndex, and my logic in remove()  method it's 2 lines, this situation is normal?